app/models/custom_field/calculated_value.rb

```diff
@@ -0,0 +1,96 @@
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+# Methods for custom fields with a format of "calculated value".
+# Should be included in the CustomField model.
+module CustomField::CalculatedValue
+  def validate_formula
+    return unless field_format_calculated_value?
+
+    if formula_string.blank?
+      errors.add(:formula, :blank)
+      return
+    end
+
+    unless formula_contains_only_allowed_characters?
+      errors.add(:formula, :invalid_characters)
+      return
+    end
+
+    # WP-64348: check for valid (i.e., visible & enabled) custom field references (see #cf_ids_used_in_formula)
+
+    # Dentaku will return nil if the formula is invalid.
+    # TODO WP-64348: add support for referenced custom fields by injecting them as variables,
+    #       e.g. Dentaku(formula_string, cf_123: CustomField.find(123).value)
+    errors.add(:formula, :invalid) unless Dentaku(formula_string)
+
+    # TODO: consider differentiating between a formula that contains missing variables, invalid
+    #       syntax, or mathematical errors.
+  end
+
+  def formula=(value)
+    if value.is_a?(String)
+      super({ formula: value, referenced_custom_fields: cf_ids_used_in_formula(value) })
+    else
+      super
+    end
+  end
+
+  # Returns the formula as a string. Will return an empty string if the formula is not set.
+  def formula_string
+    formula.is_a?(Hash) ? formula.fetch("formula", "") : formula.to_s
+  end
+
+  private
+
+  def formula_contains_only_allowed_characters?
+    # List of allowed characters in a formula. This only performs a very basic validation.
+    # Allowed characters are:
+    # + - / * ( ) whitespace digits and decimal points
+    # Additionally, the formula may contain references to custom fields in the form of `cf_123` where 123 is the ID of
+    # the custom field.
+    # Once this basic validation passes, the formula will be parsed and validated by Dentaku, which builds an AST
+    # and ensures that the formula is really valid. A welcome side effect of the basic validation done here is that
+    # it prevents built-in functions from being used in the formula, which we do not want to allow.
+    allowed_chars = %w[+ - / * ( )] + [" "]
+    split_pattern = /\s|(\+)|(-)|(\/)|(\*)|(\()|(\))/
+    allowed_tokens = /^(cf_\d+|\d+\.?\d*|\.\d+|[#{allowed_chars.join}]+)$/
```

Comment:
- `allowed_chars` are not needed here, as they can't be present after used for splitting

- `\d+\.?\d*` allows zero digits after decimal point, maybe `\d+\.?\d*|\.\d+` can be replaced with `\d*\.\d+`

- and better to use `\A…\z`

```suggestion
    allowed_tokens = /\A(cf_\d+|\d*\.\d+)\z/
```

Reply:
I think the issue with trailing decimals after the decimal point is caught by Dentaku. I have added a spec for this case.

With regards to the other suggestions:

* Used `\A\z` over `^$`
* With the usage of `Regexp.union`, the allowed chars can be omitted now, yes. In the old regex, I caught them in capture groups so they would have been visible after the split. But since we don't do anything further with the split result, it's better to use your suggestion 👍 Done! 😀